be/queue.h

```diff
@@ -0,0 +1,146 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2015-2020 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+#pragma once
+
+#ifndef __MOTR_BE_QUEUE_H__
+#define __MOTR_BE_QUEUE_H__
+
+/**
+ * @defgroup be
+ *
+ * Highlights
+ *
+ * - put()/get()/peek()/end() expect queue lock to be taken;
+ * - op callbacks are called from under m0_be_queue lock, so they MUST NOT call
+ *   m0_be_queue functions for the same bq;
+ * - queues use lists in this way: items are added at tail and are removed from
+ *   the head;
+ *
+ * Further directions
+ *
+ * - put() multiple items at once
+ *
+ * @{
+ */
+
+#include "lib/types.h"          /* m0_bcount_t */
+#include "lib/tlist.h"          /* m0_tl */
+#include "lib/mutex.h"          /* m0_mutex */
+
+#include "be/tx_credit.h"       /* m0_be_tx_credit */
+
+
+struct m0_be_op;
+struct be_queue_wait_op;
+struct m0_buf;
+
+struct m0_be_queue_cfg {
+	uint64_t    bqc_q_size_max;
+	uint64_t    bqc_producers_nr_max;
+	uint64_t    bqc_consumers_nr_max;
+	m0_bcount_t bqc_item_length;
+};
+
+struct m0_be_queue {
```

Comment:
More explanation/document is needed for this structure and its operations.
How are the items organized in the queue?
How does the get() work?
How does the put() work?

Even after reading through the be/queue.[ch], it is still very hard to understand the mechanism.

Reply:
I think multiple diagrams are needed to explain the states, or memory organization, and how the operations change the item to move from one list to another.

Reply:
Added description to the top of the file.